modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeFactory.java

```diff
@@ -266,22 +266,36 @@ else if (type instanceof UuidType)
                 return createTypeWithNullability(createSqlIntervalType(INTERVAL_QUALIFIER_DAY_TIME), true);
             else if (clazz == Period.class)
                 return createTypeWithNullability(createSqlIntervalType(INTERVAL_QUALIFIER_YEAR_MONTH), true);
-            else if (clazz == UUID.class)
-                return createTypeWithNullability(createUuidType(), true);
+            else {
+                RelDataType relType = createCustomType(clazz.getName());
+
+                if (relType != null)
+                    return relType;
+            }
         }
 
         return super.toSql(type);
     }
 
-    /** @return UUID SQL type. */
-    public RelDataType createUuidType() {
-        return canonize(new UuidType(true));
+    /** @return Custom type by name or storeable class name. {@code Null} if type not found. */
+    public RelDataType createCustomType(String typeName) {
+        return createCustomType(typeName, true);
+    }
+
+    /** @return Nullable custom type by name or storeable class name. {@code Null} if type not found. */
+    public RelDataType createCustomType(String typeName, boolean nullable) {
```

Comment:
I don't like the idea of using the same parameter in different ways (as type name and as class name), perhaps it's better to use some enum. At least we should get rid of literal "UUID" and "OTHER" and use constants.

Reply:
Changed to Type